src/meili/routing.cc

```diff
@@ -529,70 +525,75 @@ find_shortest_path(baldr::GraphReader& reader,
         break;
       }
 
+      // If this isnt the origin label then we dont need to queue any edges from it
+      // we only needed to check if it was a yet to be found destination above
+      if (destination_idx != origin_idx)
+        continue;
+
       // Expand origin: add segments from origin to destinations ahead
       // at the same edge as well as at the opposite edge to the queue
-      if (destination_idx == origin_idx) {
-        for (const auto& origin_edge : destinations[origin_idx].edges) {
-          // The tile will be guaranteed to be directededge's tile in this loop
-          const baldr::GraphTile* start_tile = nullptr;
-          const auto* directed_edge = reader.directededge(origin_edge.id, start_tile);
-
-          // Skip if edge is not allowed
-          if (!directed_edge ||
-              !IsEdgeAllowed(directed_edge, origin_edge.id, costing, label, start_tile)) {
-            continue;
-          }
+      const auto& origin = destinations[origin_idx];
+      bool allows_uturn = origin.stoptype_ == baldr::Location::StopType::BREAK ||
+                          origin.stoptype_ == baldr::Location::StopType::VIA;
+      for (const auto& origin_edge : origin.edges) {
+        // The tile will be guaranteed to be directededge's tile in this loop
+        const baldr::GraphTile* start_tile = nullptr;
+        const auto* directed_edge = reader.directededge(origin_edge.id, start_tile);
+
+        // Skip if edge is not allowed
+        if (!directed_edge ||
+            !IsEdgeAllowed(directed_edge, origin_edge.id, costing, label, start_tile)) {
+          continue;
+        }
 
-          // U-turn cost
-          float turn_cost = label.turn_cost();
-          if (label.edgeid().Is_Valid() && label.edgeid() != origin_edge.id &&
-              label.opp_local_idx() == directed_edge->localedgeidx()) {
-            turn_cost += turn_cost_table[0];
-          }
+        // Disallow immediate u-turn
```

Comment:
instead of penalizing the uturn we either allow because the type was `break` or `via` or we disallow it because it was `through` or `break_through`.